4/m_recaptchav2.cpp

```diff
@@ -0,0 +1,292 @@
+                    break;
+                }
+            }
+        }
+
+        if (!sql)
+        {
+            throw ModuleException(this, INSP_FORMAT("*** reCAPTCHA: Could not find SQL provider with id '{}'.", dbid));
+        }
+    }
+
+    ModResult OnUserPreJoin(LocalUser* user, Channel* chan, const std::string& cname, std::string& privs, const std::string& keygiven, bool override) override
+    {
+        if (user->IsOper())
+        {
+            return MOD_RES_PASSTHRU; // Oper users bypass reCAPTCHA
+        }
+
+        auto& tag = ServerInstance->Config->ConfValue("captchaconfig");
+        std::string whitelistchans = tag->getString("whitelistchans");
+        std::set<std::string> channels;
+        std::stringstream chanstream(whitelistchans);
+        std::string chname;
+        while (std::getline(chanstream, chname, ','))
+        {
+            channels.insert(chname);
+        }
+
+        if (channels.find(cname) != channels.end())
+        {
+            return MOD_RES_PASSTHRU; // Whitelisted channels bypass reCAPTCHA
+        }
+
+        std::string whitelistports = tag->getString("whitelistports");
+        std::set<int> ports;
+        std::stringstream portstream(whitelistports);
+        std::string port;
+        while (std::getline(portstream, port, ','))
+        {
+            try
+            {
+                ports.insert(std::stoi(port));
+            }
+            catch (const std::exception&)
+            {
+                ServerInstance->Logs.Warning(MODNAME, INSP_FORMAT("*** reCAPTCHA: Invalid port in whitelistports: {}", port));
+            }
+        }
+        if (ports.find(user->server_sa.port()) != ports.end())
+        {
+            return MOD_RES_PASSTHRU; // Whitelisted ports bypass reCAPTCHA
+        }
+
+        if (accountapi && accountapi->IsIdentifiedToNick(user))
+        {
+            return MOD_RES_PASSTHRU; // Identified users bypass reCAPTCHA
+        }
+
+        const std::string* status = captcha_verified.Get(user);
+        if (!status || *status != "verified")
+        {
+            std::string custom_message = tag->getString("message", "*** reCAPTCHA: You must be identified to an account or complete the reCAPTCHA before joining channels. Link: {url}");
+            size_t pos = custom_message.find("{url}");
+            if (pos != std::string::npos)
+            {
+                custom_message.replace(pos, 5, captcha_url);
+            }
+
+            user->WriteNotice(custom_message);
+
+            std::string query = INSP_FORMAT("SELECT COUNT(*) FROM ircaccess_alloweduser WHERE ip_address = '{}'", user->server_sa.addr());
+            sql->Submit(new ModuleCaptchaCheck::CheckIPQuery(this, user), query);
+
+            return MOD_RES_DENY;
+        }
+
+        return MOD_RES_PASSTHRU;
+    }
+
+    void OnUserPostInit(LocalUser* user) override
+    {
+        const std::string* status = captcha_success.Get(user);
```

Comment:
This isn't actually set anywhere? Also, you could use BoolExtItem here to reduce memory.